src/core/src/sys/new_system.rs

```diff
@@ -0,0 +1,257 @@
+// Lumol, an extensible molecular simulation engine
+// Copyright (C) 2015-2016 Lumol's contributors — BSD license
+
+//! The base type for simulation data in `lumol` is the `System` type.
+//!
+//! An `System` consists of a list of `Particle`; a list of `Molecule`
+//! specifying how the particles are bonded together; an unit cell for boundary
+//! conditions; and the interactions between these particles.
+use std::ops::{Index, IndexMut};
+use std::slice;
+use std::cmp::{min, max};
+use std::iter::IntoIterator;
+use std::i8;
+use std::collections::BTreeMap;
+
+use energy::PairInteraction;
+use energy::{BondPotential, AnglePotential, DihedralPotential};
+use types::{Vector3D, Matrix3, Zero};
+
+use super::{Particle, ParticleKind};
+use super::Molecule;
+use super::{CONNECT_12, CONNECT_13, CONNECT_14, CONNECT_FAR};
+use super::UnitCell;
+use super::interactions::Interactions;
+use super::EnergyEvaluator;
+use super::molecules::molecule_type;
+
+
+/// A particle kind. Particles with the same name will have the same kind. This
+/// is used for faster potential lookup.
+// #[derive(Clone, Copy, Hash, PartialOrd, Ord, PartialEq, Eq, Debug)]
+// pub struct ParticleKind(pub u32);
+
+#[derive(Clone)]
+pub struct Particles {
+    names: Vec<String>,
+    kinds: Vec<ParticleKind>,
+    masses: Vec<f64>,
+    charges: Vec<f64>,
+    positions: Vec<Vector3D>,
+    velocities: Vec<Vector3D>,
+    molecule_ids: Vec<usize>
+}
+
+pub struct ParticlesRef<'a> {
+    pub names: &'a [String],
+    pub kinds: &'a [ParticleKind],
+    pub masses: &'a [f64],
+    pub charges: &'a [f64],
+    pub positions: &'a [Vector3D],
+    pub velocities: &'a [Vector3D],
+    pub molecule_ids: &'a [usize]
+}
+
+pub struct ParticlesRefMut<'a> {
+    pub names: &'a mut [String],
+    pub kinds: &'a mut [ParticleKind],
+    pub masses: &'a mut [f64],
+    pub charges: &'a mut [f64],
+    pub positions: &'a mut [Vector3D],
+    pub velocities: &'a mut [Vector3D],
+    pub molecule_ids: &'a mut [usize]
+}
+
+
+#[derive(Clone)]
+pub struct SystemGeometry {
```

Comment:
Absolutely, this is just a filler name before we come up with a better one. If you think `Configuration` is better I'll go with that.